### ISSUE TITLE
[enhancement] Ability to not show restriction toast messages

### ISSUE THREAD

**Issue by Gitoffthelawn** [M66B/XPrivacy#2285]
Add the ability to not show restriction toast messages.

Ideally, this could be set for each restriction.  One possible implementation would be to long-click on a restriction, and then ask the user if they want to block toast messages for that restriction.

Alternatively, add a global option, perhaps a quirk, to not display all toast messages.

**Comment by SWADED** [comment]
+1

**Comment by kaizokan** [comment]
+1 
i definitly like the toasts but not being able to click while they are up (or cover something up) makes it not all ways desired. For some apps it not a big deal. If ever implemented my suggestion would be settings per app.

**Comment by Gitoffthelawn** [comment]
After using XPrivacy more and more, and increasing my proficiency with it, I would say this is the one enhancement that I would really like to see.  Looking at the code, I think it wouldn't involve large changes.

Thinking about this more, I see four possible implementations:
 1. A global toggle.  This would be the least flexible, and therefore the least desirable, but still would be a significant improvement.
 2. Another per-app quirk mode.  This would add more flexibility.  It would be clumsy to use, but would not require changing the UI at all.
 3. An option in the per-app settings page.  This would add more flexibility, but would require a minor UI addition.
 4. An option for each function for each app.  This would add the most flexibility and would not require changing the outward appearance of the UI at all.  One possible implementation would be to long-click on a restriction, and then ask the user if they want to block toast messages for that restriction.

Anyways, those are just my ideas.  Others may have better or different ones.

**Comment by Cerberus-tm** [comment]
I have used the Xposed module Enhanced Toast to determine that the toast message saying "Xprivacy has blocked the clipboard" came from System, not the app Xprivacy. Of course it may be that Xprivacy instructs the system to give this message.

At any rate, with the free version of Enhanced Toast, you can block any specific toast message. So, as a work-around, it is perfect and does all that I need in practice.

I, too, was annoyed with the Xprivacy clipboard message blocking the keyboard every time I opened Telegram. First I moved the toast message to the centre of the screen, instead of near the bottom, with the xposed module Xblast (Visual Tweaks → Toast Notification). That mostly solved the problem already.

**Comment by kaizokan** [comment]
Thanks good to know! even though i'm not going to use it any time soon

**Comment by Gitoffthelawn** [comment]
@Cerberus-tm Thanks. I have been thinking about trying Enhanced Toast, but people often find that system stability is inversely proportional to the number of Xposed modules installed.  Also, things tend to slow down and bugs get harder and harder to troubleshoot with more modules installed.  That said, thanks for checking it out and verifying that it works.

Even just a single toggle within XPrivacy to block all XPrivacy toasts would be useful.  There are so many things that I use XPrivacy to restrict that being notified of just some of them doesn't add any value to me (although I certainly understand and agree with Marcel's reasoning for adding them... novices may be otherwise confused).

Actually, it takes away from value because it is amazing how many apps try to do things like read the clipboard ***each and every time*** they start.  That's such a privacy violation, but mainstream apps still do it.  IIRC, even Slacker Radio reads the content of your clipboard every time it starts.  Having an XPrivacy toast message pop up every time I start some apps is a bit annoying... certainly not the end of the world or a dealbreaker... just an annoyance.

**Comment by Cerberus-tm** [comment]
@Gitoffthelawn  I understand your reasoning (although I've always happily run various Xposed modules alongside each other). If it isn't too much work, I support the idea of a new option in the per-app settings menu where you can toggle Xprivacy toasts on or off per app (I think per function is not as necessary: most of us would be fine with turning all toasts for a certain app on or off, and you'd have to create a far more complex menu if you wanted it for each function). Perhaps the code could be copied from Enchanced Toast, if its code is free.

**Comment by kaizokan** [comment]
@Gitoffthelawn I don't remember getting toasts for get clipboard permissions, but can i ask you a question about the clipboard permissions on XDA? pm me on XDA if you like. I don't know your user name and don't want to go off topic here.

**Comment by Gitoffthelawn** [comment]
@Cerberus-tm I agree that per-function would probably be overkill (although still useful in some cases).  However, I don't think it would require more visible complexity, because it could just be accessed via long-press on the function.  It would be transparent except for those that read the README.

As far as needing to copy code from Enhanced Toast to do any of this, I think it's much simpler.  A simple flag within XPrivacy can be used prevent toasts from being issues.  I think the additional code would be trivial.  I would do it myself, but I've never programmed commercially in that language, and I wouldn't want to mess up such a well-written application.

@kaizokan Can do.  Or would XMPP be better?  I've been thinking about giving that a try because Marcel recommended it, but I haven't had a chance to fire it up.

**Comment by Cerberus-tm** [comment]
@Gitoffthelawn  OK noted. I hope this will be implemented some day, but for now I'm good.
